##### AI TECH - Akademia Innowacyjnych Zastosowań Technologii Cyfrowych. Programu Operacyjnego Polska Cyfrowa na lata 2014-2020
<hr>

## <br>
<br>
<br>
<br>
<br>
<br>
<br>
<center>
    <h1 style = "font-size:48px; font-weight:normal"><b>AI</b>-TECH</h1>
    <br>
<br>
<br>
<br>
    <h1 style = "font-size:24px; font-weight:normal">Przykład 4 DO LABORATORIUM 4</h1>
    <h1 style = "font-size:24px">z UCZENIA MASZYNOWEGO</h1>
    <h1 style = "font-size:24px">Metoda gradientu prostego, perceptron</h1>
    <h1 style = "font-size:24px">Tomasz Neumann</h1>
    <h3 style = "font-size:21px">Wykorzystanie API wysokiego poziomu - Keras</h3>
</center>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
    </body>
    </html>

![stopka.svg](stopka.svg)

<center>
Projekt współfinansowany ze środków Unii Europejskiej w ramach Europejskiego Funduszu Rozwoju Regionalnego 
Program Operacyjny Polska Cyfrowa na lata 2014-2020,
Oś Priorytetowa nr 3 "Cyfrowe kompetencje społeczeństwa" Działanie  nr 3.2 "Innowacyjne rozwiązania na rzecz aktywizacji cyfrowej" 
Tytuł projektu:  „Akademia Innowacyjnych Zastosowań Technologii Cyfrowych (AI Tech)”
    </center>

<hr>

# 1. Przykład 4

W tym przykładzie wykorzystamy API wysokiego poziomu o nazwie Keras zintegrowane z pakietem TensorFlow (https://www.tensorflow.org). Zastosujemy metody optymalizacji na podstawie spadku wzdłuż gradientu.

Na początku wskażmy pakiety, które będziemy wykorzystywać w projekcie

In [1]:
import numpy as np
import tensorflow as tf

oraz wygenerujmy przykładowe, losowe dane.

In [2]:
# Parametr określający ilość danych zmiennej niezależnej x
dataset_training_size = 1000
# Prawdziwe wartości parametrów
Wt = 0.2
bt = 0.4


# Dodanie szumu do danych
# noise = 0
noise = np.random.rand(dataset_training_size).astype(np.float32)/20

# Założony model:  y = Wt* x  + bt + noise
# Generacja danych zmiennej niezależnej x
x_t = np.linspace(0, 1, dataset_training_size).astype(np.float32)

# Tu można wygenerować wartości nierównomiernie
# x_t = np.random.rand(dataset_training_size).astype(np.float32)

x_training = x_t.reshape(dataset_training_size, 1)

# Sortowanie danych, jak nie zostały stworzone liniowo
# x_training = (np.sort(x_t)).reshape(dataset_training_size,1)

W_training = np.array([Wt])

# Wyliczenie prawdziwych wartości modelu y
y_t = (np.dot(x_training, W_training.T)+bt+noise).astype(np.float32)
y_training = np.transpose(np.array([y_t]))

Definicja modelu będzie dziedziczyła po klasie Model, a następnie następuje zdefiniowanie parametrów i funkcji modelu prostej regresji liniowej.

In [ ]:
# 1. Definicja modelu
class SimpleKerasModel(tf.keras.Model):
  def __init__(self, **kwargs):
    super().__init__(**kwargs)
    self.n = 1
    self.W = self.add_weight(name="W", shape=(self.n, 1), initializer="ones", trainable=True)
    self.b = self.add_weight(name="b", shape=(1,), initializer="ones", trainable=True)

  def call(self, x, **kwargs):
    return self.W * x + self.b


Ostatecznie należy skonfigurować proces uczenia, ustawić algorytm optymalizacyjny, funkcję kosztu oraz wywołanie procedury treningu dla określonych parametrów procedury

In [6]:
# Wywołanie modelu
keras_model = SimpleKerasModel()

# Ustawienie hiperparametrów
learning_rate = .007
epochs = 100
batch_size=50

# Ustawienie parametrów treningu
keras_model.compile(
    # Użycie trybu Eager
    run_eagerly=True,
    # wybór optymalizatora
    optimizer=tf.keras.optimizers.SGD(learning_rate=learning_rate),
    # wybór funkcji straty
    loss="mse"
)

# 5. Trening modelu z ustawionymi hiperparametrami
keras_model.fit(x_training, y_training, epochs=epochs, batch_size=batch_size)

print("\nEstimated W: ", keras_model.W.numpy())
print("\nEstimated b: ", keras_model.b.numpy())

Epoch 1/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 1s 22ms/step - loss: 0.7340
Epoch 2/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 21ms/step - loss: 0.3646
Epoch 3/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 21ms/step - loss: 0.1838
Epoch 4/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 20ms/step - loss: 0.0953
Epoch 5/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 18ms/step - loss: 0.0518
Epoch 6/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 18ms/step - loss: 0.0304
Epoch 7/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 1s 21ms/step - loss: 0.0197
Epoch 8/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 17ms/step - loss: 0.0144
Epoch 9/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 19ms/step - loss: 0.0116
Epoch 10/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 19ms/step - loss: 0.0101
Epoch 11/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 19ms/step - loss: 0.0092
Epoch 12/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 21ms/step - loss: 0.0086
Epoch 13/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 21ms/step - loss: 0.0082
Epoch 14/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - loss: 0.0078
Epoch 15/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 17ms/step - lo

co pozwala już na analizę uzyskanych wyników.

# Zadania do wykonania

<p><b>Zadanie 1</b></p>
Przeprowadź eksperymenty dla różnych wartości batch-size oraz współczynnika uczenia. Analizy przeprowadź na podstawie funkcji straty w zależności od numeru epoki.


In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import tensorflow as tf


def make_training_data(seed=42):
    rng = np.random.default_rng(seed)
    sample_count = 1000
    true_w = 0.2
    true_b = 0.4
    noise = rng.random(sample_count).astype(np.float32) / 20
    x_values = np.linspace(0, 1, sample_count).astype(np.float32)
    x_train = x_values.reshape(sample_count, 1)
    y_train = (x_train * true_w + true_b + noise.reshape(-1, 1)).astype(np.float32)
    return x_train, y_train


def build_model():
    return tf.keras.Sequential([
        tf.keras.layers.Input(shape=(1,)),
        tf.keras.layers.Dense(
            1,
            kernel_initializer=tf.keras.initializers.Ones(),
            bias_initializer=tf.keras.initializers.Ones(),
        ),
    ])


def run_experiment(batch_size, learning_rate, epochs=30):
    tf.keras.backend.clear_session()
    tf.random.set_seed(42)
    x_train, y_train = make_training_data()
    model = build_model()
    model.compile(optimizer=tf.keras.optimizers.SGD(learning_rate=learning_rate), loss="mse")
    history = model.fit(x_train, y_train, epochs=epochs, batch_size=batch_size, verbose=0, shuffle=False)
    losses = history.history["loss"]
    weights, bias = model.layers[0].get_weights()
    return losses, weights.ravel()[0], bias.ravel()[0]


configs = [
    (10, 0.001),
    (10, 0.007),
    (50, 0.007),
    (100, 0.007),
    (50, 0.05),
]

rows = []
for batch_size, learning_rate in configs:
    losses, weight, bias = run_experiment(batch_size, learning_rate)
    label = f"batch={batch_size}, lr={learning_rate}"
    plt.plot(losses, label=label)
    rows.append({
        "batch_size": batch_size,
        "learning_rate": learning_rate,
        "first_loss": losses[0],
        "final_loss": losses[-1],
        "W": weight,
        "b": bias,
    })

plt.xlabel("epoch")
plt.ylabel("loss")
plt.legend()
plt.show()

results = pd.DataFrame(rows)
print(results.to_string(index=False, float_format=lambda x: f"{x:.6f}"))


 batch_size  learning_rate  first_loss  final_loss        W        b
         10       0.001000    0.791146    0.006312 0.464339 0.278180
         10       0.007000    0.263077    0.000313 0.231443 0.403470
         50       0.007000    0.732590    0.004838 0.426697 0.296558
        100       0.007000    0.862853    0.007822 0.496052 0.266475
         50       0.050000    0.187146    0.000226 0.210525 0.417037


<p><b>Wnioski do zadania 1</b></p>

Najmniejszą końcową stratę uzyskał wariant `batch_size=50`, `learning_rate=0.05`: `final_loss=0.000226`, przy parametrach `W=0.210` i `b=0.417`. Bardzo podobny wynik dał `batch_size=10`, `learning_rate=0.007` (`final_loss=0.000313`, `W=0.231443`, `b=0.403470`).

Przy tym samym `learning_rate=0.007` większy `batch_size` pogarszał wynik po 30 epokach: dla `batch_size=10` strata końcowa wyniosła `0.000313`, dla `50` już `0.004838`, a dla `100` `0.007822`. Z kolei zbyt mały `learning_rate=0.001` uczył się wolniej (`final_loss=0.006312`). Na podstawie tych wyników najlepszy z testowanych wariantów to `batch_size=50` i `learning_rate=0.05`.

Najbardziej stabilne i sensowne wyniki pochodzą z ustawień, w których błąd maleje, ale parametry nie eksplodują. Dlatego wnioski powinny opierać się na konkretnych wartościach metryk i przebiegach straty, a nie na ogólnym założeniu, że większy learning rate jest zawsze lepszy albo że większy batch_size jest zawsze lepszy i tego typu. Często te parametry trzeba dobrać do konkretnego problemu, więc trudno tutaj pisać konkretne zalecenia czy konkretne wartości tych parametrów.
